0.preprocess-sites/7.visualize-cell-summary.py

```diff
@@ -0,0 +1,249 @@
+barcode_cols = spot_args["barcode_cols"]
+barcode_cols = ["Metadata_Foci_" + col for col in barcode_cols]
+gene_cols = spot_args["gene_cols"]
+gene_cols = ["Metadata_Foci_" + col for col in gene_cols]
+spot_score_cols = spot_args["spot_score_cols"]
+spot_score_count_cols = ["Metadata_Foci_" + col + "_count" for col in spot_score_cols]
+spot_score_mean_cols = ["Metadata_Foci_" + col + "_mean" for col in spot_score_cols]
+
+input_basedir = cell_args["output_basedir"]
+metadata_foci_col = cell_args["metadata_merge_columns"]["cell_quality_col"]
+cell_cols = cell_args["metadata_merge_columns"]["cell_cols"]
+cell_quality_col = cell_args["metadata_merge_columns"]["cell_quality_col"]
+foci_site_col = cell_args["foci_site_col"]
+
+output_resultsdir = summ_cell_args["output_resultsdir"]
+output_figuresdir = summ_cell_args["output_figuresdir"]
+cell_category_order = summ_cell_args["cell_category_order"]
+cell_category_colors = summ_cell_args["cell_category_colors"]
+
+cell_quality = CellQuality(quality_func)
+cell_category_dict = cell_quality.define_cell_quality()
+empty_cell_category = len(cell_category_dict) + 1
+cell_category_dict[empty_cell_category] = "Empty"
+cell_category_df = pd.DataFrame(cell_category_dict, index=["Cell_Class"])
+cell_category_list = list(cell_category_dict.values())
+
+# Read and Merge Data
+cell_quality_list = []
+metadata_list = []
+metadata_col_list = []
+metadata_col_list.append("Cell_Class")
+metadata_col_list.extend(cell_cols)
+metadata_col_list.extend(barcode_cols)
+metadata_col_list.extend(gene_cols)
+metadata_col_list.extend(spot_score_count_cols)
+metadata_col_list.extend(spot_score_mean_cols)
+metadata_col_list.append(cell_quality_col)
+metadata_col_list.append(foci_site_col)
+
+input_dir = pathlib.Path(input_basedir, batch, "paint")
+sites = os.listdir(input_dir)
+print(f"There are {len(sites)} sites.")
+
+for site in sites:
+
+    cell_count_file = pathlib.Path(input_dir, site, f"cell_counts_{site}.tsv")
+    metadata_file = pathlib.Path(input_dir, site, f"metadata_{site}.tsv.gz")
+
+    # Aggregates cell quality by site into single list
+    cell_quality_list.append(pd.read_csv(cell_count_file, sep='\t'))
+
+    # Aggregates metadata by site into a single list
+    metadata_df = (
+        pd
+        .read_csv(metadata_file, sep='\t')
+        .loc[:,metadata_col_list]
+        .reset_index(drop=True)
+    )
+
+    metadata_list.append(metadata_df)
+
+# Creates dataframe from cell quality list
+cell_count_df = pd.concat(cell_quality_list, axis="rows").reset_index(drop=True)
+
+# Assigns the Cell_Quality column to the category datatype and sets categories
+cell_count_df.loc[:, 'Cell_Quality'] = (
+    pd.Categorical(
+        cell_count_df.Cell_Quality,
+        categories=cell_category_order
+    )
+)
+
+# Assigns the Site column to the category datatype
+cell_count_df.loc[:, "site"] = (
+    pd.Categorical(
+        cell_count_df.site,
+        categories=(
+            cell_count_df
+            .groupby("site")
```

Comment:
1) I don't fully understand what all of the steps here are doing to the site column
2) I'm having problems with .groupby (here and elsewhere). I believe I'm understanding that this should sort the sites in descending order of the sum of their cell_count, but it's not working.

Reply:
the `pd.Categorical()` command is the key thing. It is a strange (but powerful) nuance that is important for plotting purposes. With `pd.Categorical()` one can define the order of an axis containing categorical variables.

So the understanding in 2 is correct. What error are we getting with `.groupby`?

Reply:
It doesn't throw an error but from what I can tell it's just not actually doing anything.